File: src/Experiments/Benchmarks/TypeDispatch.CSharp/TypeDispatch.cs
using System.Numerics;

namespace TypeDispatch.CSharp;

public class TypeDispatch
{
    public static bool IsNumeric(object o)
    {
        return o != null && IsNumericType(o.GetType());
    }

    public static bool IsNumericType(Type type)
    {
        switch (Type.GetTypeCode(type))
        {
            case TypeCode.SByte:
            case TypeCode.Byte:
            case TypeCode.Int16:
            case TypeCode.Int32:
            case TypeCode.Int64:
            case TypeCode.Double:
            case TypeCode.Single:
            case TypeCode.UInt16:
            case TypeCode.UInt32:
            case TypeCode.UInt64:
                return true;
            default: 
                return type == typeof(BigInteger);
        }
        //if (type == typeof(BigInteger) )
        //    return true;
        //return false;
    }

    public static bool HasSpecialType(Type t)
    {
        return t == typeof(BigInteger) || t == typeof(String) || t == typeof(DateTime) || t == typeof(Uri);
    }

    public static bool TestObjectType(object o)
    {
        return HasSpecialType(o.GetType()); 
    }

}
    

File: src/Experiments/Benchmarks/IsNumericTest/TypeDispatch.fs

module TypeDispatch

open System

type TypeDispatch() =

    static member public IsNumericType(t: Type) =
        match Type.GetTypeCode(t) with
        | TypeCode.SByte
        | TypeCode.Byte
        | TypeCode.Int16
        | TypeCode.Int32
        | TypeCode.Int64
        | TypeCode.Double
        | TypeCode.Single
        | TypeCode.UInt16
        | TypeCode.UInt32
        | TypeCode.UInt64 -> true
        | _ -> System.Type.op_Equality(t, typeof<System.Numerics.BigInteger>)

    static member IsNumeric(o: obj) =
        match o with
        | null -> false
        | _ -> TypeDispatch.IsNumericType(o.GetType())
